src/persist-client/src/operators/shard_source.rs

```diff
@@ -593,6 +626,126 @@ where
     (descs_stream, activate_on_drop)
 }
 
+pub(crate) fn granular_backpressure<T, G>(
+    scope: &G,
+    name: &str,
+    descs: &Stream<G, (usize, SerdeLeasedBatchPart)>,
+    flow_control: GranularFlowControl<G>,
+) -> (Stream<G, (usize, SerdeLeasedBatchPart)>, Rc<dyn Any>)
+where
+    T: Timestamp + Lattice + Codec64,
+    G: Scope<Timestamp = Hybrid<T, u64>>,
+{
+    let (flow_control_stream, flow_control_max_bytes) = (
+        flow_control.progress_stream,
+        flow_control.max_inflight_bytes,
+    );
+
+    let mut builder = AsyncOperatorBuilder::new(
+        format!("shard_source_granular_backpressure({})", name),
+        scope.clone(),
+    );
+    let (mut descs_output, descs_stream) = builder.new_output();
+
+    let mut descs_input = builder.new_input(descs, Pipeline);
+    let mut flow_control_input =
+        builder.new_input_connection(&flow_control_stream, Pipeline, vec![Antichain::new()]);
+
+    let mut inflight_parts = Vec::new();
+
+    let descs_input = async_stream::stream!({
+        let mut part_number = 0;
+        let mut pending_parts = Vec::new();
+        loop {
+            for part in pending_parts.drain(..) {
+                yield part;
+            }
+
+            match descs_input.next_mut().await {
+                None => break,
+                Some(Event::Data(cap, data)) => {
+                    for d in data.drain(..) {
+                        // Refine the timestamp to the part number.
+                        let mut time = cap.time().clone();
+                        time.1 = part_number;
+                        pending_parts.push((cap.delayed(&time), d));
```

Comment:
Why do we need to stage these in `pending_parts` and not `yield` in here directly?

Reply:
This is a good point!